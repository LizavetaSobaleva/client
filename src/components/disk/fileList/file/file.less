@import '../../../../utils/colors.less';

.file {
    padding: 8px;
    border-bottom: 1px solid @lightgrey;
    display: grid;
    grid-template-columns: 0.5fr 5fr repeat(2, 1fr) 0.5fr;
    color: @subtext;
    font-weight: 400;
    align-items: center;
    transition: ease-in 50ms all;

    &:hover {
        background-color: #f9fafc;
    }

    &__img {
        height: 40px;
        justify-self: center;
    }

    &__name {
        grid-column-start: 2;
        width: 98%;
        white-space: nowrap;
        overflow: hidden;
        text-overflow: ellipsis;
    }

    &__date {
        grid-column-start: 3;
        justify-self: center;
    }

    &__size {
        grid-column-start: 4;
        justify-self: center;
    }

    &__actions {
        grid-column-start: 5;
        justify-self: center;
        position: relative;
    }

    &__menu {
        height: 20px;
        width: 23px;
    }

    &__options {
        display: none;
        position: absolute;
        top: 50%;
        right: 50%;
        box-shadow: 2px 2px 6px @grey;
        border-radius: 8px;

        .popup {
            padding: 4px;

            &__header {
                display: none;
            }

            button {
                width: 100%;
                border-radius: 4px;
                margin: 0;
            }
        }
    }

    &__actions:hover &__options {
        display: block;
    }
}

.file-plate {
    width: 150px;
    height: 150px;
    margin: 4px;
    padding: 16px 12px;
    border: 1px solid @borders;
    border-radius: 8px;
    color: @subtext;
    font-weight: 400;
    display: flex;
    flex-direction: column;
    align-items: center;
    transition: ease-in 50ms all;
    position: relative;

    &:hover {
        background-color: #f9fafc;
    }

    &__img {
        height: 60px;
        margin: 8px 0;
    }

    &__name {
        margin-bottom: 4px;
        max-width: 98%;
        white-space: nowrap;
        overflow: hidden;
        text-overflow: ellipsis;
    }

    &__size {
        font-size: 12px;
        color: @subtext-80;
        letter-spacing: -0.4px;
    }

    &__date {
        display: none;
    }

    &__actions {
        position: absolute;
        top: 0;
        right: 0;
    }

    &__menu {
        height: 20px;
        width: 23px;
    }

    &__options {
        display: none;
        position: absolute;
        top: 50%;
        right: 50%;
        box-shadow: 2px 2px 6px @grey;
        border-radius: 8px;

        .popup {
            padding: 4px;

            &__header {
                display: none;
            }

            button {
                width: 100%;
                border-radius: 4px;
                margin: 0;
            }
        }
    }

    &__actions:hover &__options {
        display: block;
    }
}

@media only screen and (max-width: 660px){
    .file {

        &__date {
            display: none
        }
    
        &__size {
            display: none
        }
    }
}