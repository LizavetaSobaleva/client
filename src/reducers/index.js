import { applyMiddleware, combineReducers } from "redux";
import { configureStore} from "@reduxjs/toolkit"
import { thunk } from "redux-thunk";
import userReducer from "./userReducer";
import fileReducer from "./fileReducer";
import uploadReducer from "./uploadReducer";
import appReducer from "./appReducer";
import { composeWithDevTools } from '@redux-devtools/extension';


const rootReducer = combineReducers({
    user: userReducer,
    files: fileReducer,
    upload: uploadReducer,
    app: appReducer
})

export const store = configureStore({reducer: rootReducer, devTools: composeWithDevTools(applyMiddleware(thunk))})