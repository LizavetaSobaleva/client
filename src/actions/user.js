import axios from "axios"
import { setUser } from '../reducers/userReducer'


export const instanceAxios = axios.create({
     baseURL: process.env.REACT_APP_API_URL + "/api/"
})

export const registration = async (name, email, password, dispatch) => {
   try {
        const response = await instanceAxios.post('auth/registration', {
            name,
            email,
            password
        })
        alert(response.data.message)
        dispatch(login(email, password))
   } catch (e) {
        alert(e.response.data.message)
   }
}

export const login = (email, password) => {
     return async dispatch => {
          try {
               const response = await instanceAxios.post('auth/login', {
                   email,
                   password
               })
               dispatch(setUser(response.data.user))
               localStorage.setItem('token', response.data.token)
          } catch (e) {
               alert(e.response.data.message)
          }
     }
}

export const auth = () => {
     return async dispatch => {
          try {
               const response = await instanceAxios.get('auth/auth',
                    { headers: { Authorization: `Bearer ${localStorage.getItem('token')}` }}
               )
               localStorage.setItem('token', response.data.token)
               dispatch(setUser(response.data.user))
          } catch (e) {
               console.log(e.response.data.message)
               localStorage.removeItem('token')
          }
     }
}

export const uploadAvatar = (file) => {
     return async dispatch => {
          try {
               const formData = new FormData()
               formData.append('file', file)
               const response = await instanceAxios.post('files/avatar', formData,
                    { headers: { Authorization: `Bearer ${localStorage.getItem('token')}` }}
               )
               dispatch(setUser(response.data))
          } catch (e) {
               console.log(e)
          }
     }
}

export const deleteAvatar = () => {
     return async dispatch => {
          try {
               const response = await instanceAxios.delete('files/avatar',
                    { headers: { Authorization: `Bearer ${localStorage.getItem('token')}` }}
               )
               dispatch(setUser(response.data))
          } catch (e) {
               console.log(e)
          }
     }
}