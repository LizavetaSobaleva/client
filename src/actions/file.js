import { hideLoader, showLoader } from "../reducers/appReducer"
import { addFile, deleteFileAction, setFiles, setStructure } from "../reducers/fileReducer"
import { addUploadFile, changeUploadFile, showUploader } from "../reducers/uploadReducer"
import { instanceAxios } from "./user"

export function getFiles(dirId) {
  return async (dispatch) => {
    try {
      dispatch(showLoader())
      const response = await instanceAxios.get(`files${dirId ? '?parent='+dirId : ''}`,
        { headers: { Authorization: `Bearer ${localStorage.getItem('token')}` }}
      )
      dispatch(setFiles(response.data))
    } catch (e) {
      console.log(e.response.data.message)
    } finally {
      dispatch(hideLoader())
    }
  }
}

export function createDir(dirId, name) {
  return async (dispatch) => {
    try {
      const response = await instanceAxios.post(`files`, {
          name,
          parent: dirId,
          type: 'dir'
        },
        { headers: { Authorization: `Bearer ${localStorage.getItem('token')}` }}
        )
      dispatch(addFile(response.data))
    } catch (e) {
      alert(e.response.data.message)
    }
  }
} 

export function uploadFile(file, dirId) {
  return async (dispatch) => {
    try {
      const formData = new FormData()
      formData.append('file', file)
      if(dirId) {
        formData.append('parent', dirId)
      }
      const uploadFile = {name: file.name, process: 0, id: Date.now()}
      dispatch(showUploader())
      dispatch(addUploadFile(uploadFile))
      const response = await instanceAxios.post(`files/upload`, formData, {
                headers: {Authorization: `Bearer ${localStorage.getItem('token')}`},
                onUploadProgress: progressEvent => {
            const totalLength = progressEvent.event.lengthComputable ? progressEvent.event.total : progressEvent.event.target.getResponseHeader('content-length') || progressEvent.event.target.getResponseHeader('x-decompressed-content-length')
            if(totalLength) {
              uploadFile.progress = Math.round((progressEvent.loaded * 100) / totalLength)
              dispatch(changeUploadFile(uploadFile))
            }
          }
        }
      )
      dispatch(addFile(response.data))
    } catch (e) {
      alert(e.response.data.message)
    }
  }
} 

export async function downloadFile(file) {
  try {
    const response = await instanceAxios.get(`files/download?id=${file._id}`, {
      responseType: 'blob',
      headers: { Authorization: `Bearer ${localStorage.getItem('token')}` }}
    )
    const blob = new Blob([response.data], { type: response.headers['content-type']})

    const downloadUrl = window.URL.createObjectURL(blob)
    const link = document.createElement('a')
    link.href = downloadUrl
    link.download = file.name

    document.body.appendChild(link)
    link.click()
    link.remove()
  } catch (e) {
    console.log(e.response.data.message)
  }
}

export function deleteFile(file) {
  return async (dispatch) => {
    try {
      const response = await instanceAxios.delete(`files?id=${file._id}`,
        { headers: { Authorization: `Bearer ${localStorage.getItem('token')}` }}
      )
      dispatch(deleteFileAction(file._id))
      alert(response.data.message)
    } catch (e) {
      alert(e.response.data.message)
    }
  }
}

export function searchFiles(search) {
  return async (dispatch) => {
    try {
      const response = await instanceAxios.get(`files/search?search=${search}`,
        { headers: { Authorization: `Bearer ${localStorage.getItem('token')}` }}
      )
      dispatch(setFiles(response.data))
    } catch (e) {
      alert(e?.response?.data?.message)
    } finally {
      dispatch(hideLoader())
    }
  }
}

export function getStructure() {
  return async (dispatch) => {
    try {
      const response = await instanceAxios.get(`files/structure`,
        { headers: { Authorization: `Bearer ${localStorage.getItem('token')}` }}
      )
      dispatch(setStructure(response.data))
    } catch (e) {
      alert(e?.response?.data?.message)
    } finally {
      dispatch(hideLoader())
    }
  }
}