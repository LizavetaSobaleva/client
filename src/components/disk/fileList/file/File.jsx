import React from 'react'
import './file.less'
import dirLogo from '../../../../assets/img/folder.svg'
import fileLogo from '../../../../assets/img/file.svg'
import { useDispatch, useSelector } from 'react-redux'
import { pushToStack, setCurrentDir } from '../../../../reducers/fileReducer'
import SecondaryButton from '../../../UI/secondaryButton/SecondaryButton'
import options from '../../../../assets/img/options.svg'
import Popup from '../../../UI/popup/Popup'
import { deleteFile, downloadFile } from '../../../../actions/file'
import sizeFormat from '../../../../utils/sizeFormat'

const File = ({file}) => {
    const dispatch = useDispatch()
    const currentDir = useSelector(state => state.files.currentDir)
    const fileView = useSelector(state => state.app.view)

    function openHandler() {
      if(file.type === 'dir') {
        dispatch(pushToStack(currentDir))
        dispatch(setCurrentDir(file._id))
      }
      }

    function downloadClickHandler(e) {
      e.stopPropagation()
      downloadFile(file)
    }

    // function renameClickHandler(e) {
    // }

    function deleteClickHandler(e) {
      e.stopPropagation()
      dispatch(deleteFile(file))
    }


  if (fileView === 'list') {
    return (
      <div className='file' onClick={() => openHandler(file)}>
        <img src={file.type === 'dir' ? dirLogo : fileLogo} alt="" className='file__img'/>
        <div className="file__name">{file.name}</div>
        <div className="file__date">{file.date.slice(0, 10)}</div>
        <div className="file__size">{file.size !== 0 && sizeFormat(file.size)}</div>
        <div className="file__actions">
          <SecondaryButton style={{padding: 8}}>
            <img src={options} alt="" className='file__menu'/>
          </SecondaryButton>
          <div className="file__options">
            <Popup>
              {file.type !== 'dir' && <SecondaryButton onClick={(e) => downloadClickHandler(e)}>Download</SecondaryButton>}
              {/* <SecondaryButton onClick={(e) => renameClickHandler(e)}>Rename</SecondaryButton> */}
              <SecondaryButton onClick={(e) => deleteClickHandler(e)}>Delete</SecondaryButton>
            </Popup>
            </div>
        </div>
      </div>
    )
  }

  if (fileView === 'plate') {
    return (
      <div className='file-plate' onClick={() => openHandler(file)}>
          <img src={file.type === 'dir' ? dirLogo : fileLogo} alt="" className='file-plate__img'/>
          <div className="file-plate__name">{file.name}</div>
          <div className="file-plate__date">{file.date.slice(0, 10)}</div>
          <div className="file-plate__size">{file.size !== 0 && sizeFormat(file.size)}</div>
          <div className="file-plate__actions">
            <SecondaryButton style={{padding: 8}}>
              <img src={options} alt="" className='file-plate__menu'/>
            </SecondaryButton>
              <div className="file-plate__options">
                <Popup>
                  {file.type !== 'dir' && <SecondaryButton onClick={(e) => downloadClickHandler(e)}>Download</SecondaryButton>}
                  {/* <SecondaryButton onClick={(e) => renameClickHandler(e)}>Rename</SecondaryButton> */}
                  <SecondaryButton onClick={(e) => deleteClickHandler(e)}>Delete</SecondaryButton>
                </Popup>
              </div>
          </div>
      </div>
    )
  }
}

export default File