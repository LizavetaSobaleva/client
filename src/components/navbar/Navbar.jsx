import React, { useState } from 'react'
import './navbar.less'
import Logo from '../../assets/img/logo.png'
import { NavLink } from 'react-router-dom'
import { useDispatch, useSelector } from 'react-redux'
import Input from '../UI/input/Input'
import { getFiles, searchFiles } from '../../actions/file'
import avatarLogo from '../../assets/img/user_primary.svg'
import { showLoader } from '../../reducers/appReducer'

const Navbar = () => {
  const isAuth = useSelector(state => state.user.isAuth)
  const currentDir = useSelector(state => state.files.currentDir)
  const currentUser = useSelector(state => state.user.currentUser)
  const dispatch = useDispatch()
  const [searchName, setSearchName] = useState('')
  const [searchTimout, setSearchTimout] = useState(false)
  const avatar = currentUser.avatar ? `${process.env.REACT_APP_API_URL}/${currentUser.avatar}` : avatarLogo


  function searchChangeHandler(e) {
    setSearchName(e.target.value)
    if(searchTimout !== false) {
      clearTimeout(searchTimout)
    }
    dispatch(showLoader())
    if(e.target.value !== '') {
      setSearchTimout(setTimeout((value) => {
        dispatch(searchFiles(value))
      }, 500, e.target.value))
    } else {
      dispatch(getFiles(currentDir))
    }
  }
  

  return (
    <div className='navbar' data-testid="navbar">
      <div className="navbar__content" data-testid="navbarContent">
      {!isAuth ? 
        <>
          <div className="navbar__header" data-testid="navbarHeader">
            <img src={Logo} alt="" className="navbar__logo" data-testid="navbarLogo"/>
            <span>MERN cloud</span>
          </div>
          <div className="navbar__btns" data-testid="navbarButtons">
            <div className="navbar__login" data-testid="loginButton"><NavLink to="/login">Log in</NavLink></div>
            <div className="navbar__registration" data-testid="registerButton"><NavLink to="/registration">Sign up</NavLink></div>
          </div>
        </>
        :
        <>
          <div className="navbar__header" data-testid="searchHeader">
            <Input type='search' placeholder="Type to search..."
              value={searchName}
              onChange={e => searchChangeHandler(e)}
              data-testid="searchInput"
            />
          </div>
          <div className="navbar__user" data-testid="navbarUser">
            <div className="navbar__message" data-testid="userMessage">Hello, {currentUser.name}!</div>
            <NavLink to="/profile">
              <img src={avatar} alt="" className="navbar__avatar" data-testid="userAvatar"/>
            </NavLink>
          </div>
        </>
      }
        </div>
    </div>
  )
}

export default Navbar