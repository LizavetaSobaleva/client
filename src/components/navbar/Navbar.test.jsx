import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import { MemoryRouter } from 'react-router-dom';
import Navbar from './Navbar';
import { getFiles, searchFiles } from '../../actions/file';
import { showLoader } from '../../reducers/appReducer';

jest.mock('../../actions/file', () => ({
  getFiles: jest.fn().mockReturnValue({ type: 'SET_FILES', type: 'SET_LOADER' }),
  searchFiles: jest.fn().mockReturnValue({ type: 'SET_FILES', type: 'SET_LOADER' }),
}));

jest.mock('../../reducers/appReducer', () => ({
  showLoader: jest.fn().mockReturnValue({ type: 'SET_LOADER' }),
}));

const mockStore = configureStore([]);
let store;

const renderComponent = (store) =>
  render(
    <Provider store={store}>
      <MemoryRouter>
        <Navbar />
      </MemoryRouter>
    </Provider>
  );

describe('Navbar Component', () => {
  beforeEach(() => {
    store = mockStore({
      user: { isAuth: false, currentUser: { name: 'John Doe', avatar: null } },
      files: { currentDir: 'root' },
    });
    store.dispatch = jest.fn(); // Mock the store.dispatch function
  });

  test('renders Navbar without crashing', () => {
    const { asFragment } = renderComponent(store);
    expect(screen.getByTestId('navbar')).toBeInTheDocument();
    expect(screen.getByTestId('navbarContent')).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });

  test('renders login and register buttons when user is not authenticated', () => {
    const { asFragment } = renderComponent(store);
    expect(screen.getByTestId('loginButtonLink')).toBeInTheDocument();
    expect(screen.getByTestId('registerButtonLink')).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });

  test('renders search input and user info when user is authenticated', () => {
    store = mockStore({
      user: { isAuth: true, currentUser: { name: 'John Doe', avatar: null } },
      files: { currentDir: 'root' },
    });
    const { asFragment } = renderComponent(store);
    expect(screen.getByTestId('searchInput')).toBeInTheDocument();
    expect(screen.getByTestId('userMessage')).toHaveTextContent('Hello, John Doe!');
    expect(asFragment()).toMatchSnapshot();
  });

  test('search input triggers searchFiles action', () => {
    store = mockStore({
      user: { isAuth: true, currentUser: { name: 'John Doe', avatar: null } },
      files: { currentDir: 'root' },
    });
    store.dispatch = jest.fn(); // Mock the store.dispatch function

    renderComponent(store);
    fireEvent.change(screen.getByTestId('searchInput'), { target: { value: 'test' } });
    expect(store.dispatch).toHaveBeenCalledWith(showLoader());

    jest.useFakeTimers();
    fireEvent.change(screen.getByTestId('searchInput'), { target: { value: 'test' } });
    jest.advanceTimersByTime(500);
    expect(store.dispatch).toHaveBeenCalledWith(searchFiles('test'));
    jest.useRealTimers();
  });

  test('search input triggers getFiles action when cleared', () => {
    store = mockStore({
      user: { isAuth: true, currentUser: { name: 'John Doe', avatar: null } },
      files: { currentDir: 'root' },
    });
    store.dispatch = jest.fn(); // Mock the store.dispatch function

    renderComponent(store);
    fireEvent.change(screen.getByTestId('searchInput'), { target: { value: 'test' } });
    fireEvent.change(screen.getByTestId('searchInput'), { target: { value: '' } });
    expect(store.dispatch).toHaveBeenCalledWith(getFiles('root'));
  });
});
