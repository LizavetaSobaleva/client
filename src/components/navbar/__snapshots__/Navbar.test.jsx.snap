// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Navbar Component renders Navbar without crashing 1`] = `
<DocumentFragment>
  <div
    class="navbar"
    data-testid="navbar"
  >
    <div
      class="navbar__content"
      data-testid="navbarContent"
    >
      <div
        class="navbar__header"
        data-testid="navbarHeader"
      >
        <img
          alt=""
          class="navbar__logo"
          data-testid="navbarLogo"
          src="test-file-stub"
        />
        <span>
          MERN cloud
        </span>
      </div>
      <div
        class="navbar__btns"
        data-testid="navbarButtons"
      >
        <div
          class="navbar__login"
          data-testid="loginButtonLink"
        >
          <a
            class=""
            href="/login"
          >
            Log in
          </a>
        </div>
        <div
          class="navbar__registration"
          data-testid="registerButtonLink"
        >
          <a
            class=""
            href="/registration"
          >
            Sign up
          </a>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Navbar Component renders login and register buttons when user is not authenticated 1`] = `
<DocumentFragment>
  <div
    class="navbar"
    data-testid="navbar"
  >
    <div
      class="navbar__content"
      data-testid="navbarContent"
    >
      <div
        class="navbar__header"
        data-testid="navbarHeader"
      >
        <img
          alt=""
          class="navbar__logo"
          data-testid="navbarLogo"
          src="test-file-stub"
        />
        <span>
          MERN cloud
        </span>
      </div>
      <div
        class="navbar__btns"
        data-testid="navbarButtons"
      >
        <div
          class="navbar__login"
          data-testid="loginButtonLink"
        >
          <a
            class=""
            href="/login"
          >
            Log in
          </a>
        </div>
        <div
          class="navbar__registration"
          data-testid="registerButtonLink"
        >
          <a
            class=""
            href="/registration"
          >
            Sign up
          </a>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Navbar Component renders search input and user info when user is authenticated 1`] = `
<DocumentFragment>
  <div
    class="navbar"
    data-testid="navbar"
  >
    <div
      class="navbar__content"
      data-testid="navbarContent"
    >
      <div
        class="navbar__header"
        data-testid="searchHeader"
      >
        <input
          data-testid="searchInput"
          placeholder="Type to search..."
          type="search"
          value=""
        />
      </div>
      <div
        class="navbar__user"
        data-testid="navbarUser"
      >
        <div
          class="navbar__message"
          data-testid="userMessage"
        >
          Hello, John Doe!
        </div>
        <a
          class=""
          href="/profile"
        >
          <img
            alt=""
            class="navbar__avatar"
            data-testid="userAvatar"
            src="test-file-stub"
          />
        </a>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
